### ISSUE TITLE
Rule proposal: No property references in destructuring patterns

### ISSUE THREAD

**Issue by brettz9** [sindresorhus/eslint-plugin-unicorn#1660]
### Description

While using property references in destructuring patterns can be expressive, it can suffer from two problems:

1. It is confuseable with destructuring default values, e.g., `({ a: this.b } = obj);` and `({ a = this.b } = obj);`

2. Tracing the however intended side effects may be confusing as one might be inclined to gloss over full property references on the left hand side where they don't normally occur and where normally only some straightforward local declarations occur.

3. Property references on the left-hand side may allow extended complex and difficult to read expressions:
```js
var a = true, b = {d: 4}, c = {d: 88};
({x: (a ? b : c).d} = {x: 5}); // b.d === 5
b = false;
({x: (a ? b : c).d} = {x: 5}); // c.d === 5
```
While local variables are also defined within the left-hand-side, these are only local in scope and easier to visually track).

4. Left-hand property reference expressions as above may have been more likely intended for the right side.

### Fail

```js
var obj = {};
({ a: this.b } = obj);
```
```js
var obj = {};
({ a: (true).b } = obj);
```
```js
var arr = [];
([ this.a ] = arr);
```

### Pass

```js
var c;
var obj = {a: 5};
({ a: c } = obj);
```
```js
var obj = {a: {b: 15}};
({ a: {b: {c} }} = obj);
```
```js
var a;
var arr = [3];
[ a ] = arr;
```

**Comment by fisker** [comment]
From my experience

1. Only assign to variable

```js
// Good
({a} = v)
({a: renamed} = v)
// Bad
({a: a.b} = v)
```

```js
// Good
[a] = v;
// Bad
[a.b] = v;
```

2. Assign should not use default value

```js
// Bad
({a = defaultValue} = v);
```

```js
// Bad
[a = defaultValue] = v;
```

2. No deep destructuring assign

```js
// Bad
({a: {b}} = v);
({a: [b]} = v);
```

```js
// Bad
[[a]] = v;
[{a}] = v;
```

3. No default value and rename in deep destructuring 

```js
// Good
const {a} = v;
const [a] = v
const {a: renamed = defaultValue} = v;
// Bad
const {a: {b = defaultValue}} = v;
const {a: {b: renamed}} = v;
const {a: {b: renamed = defaultValue}} = v;
```

**Comment by brettz9** [comment]
Yeah maybe some boolean options to toggle between allowing or disallowing any combination of that granularity should be especially appreciated by projects.

**Comment by fisker** [comment]
`no-restricted-destructuring-patterns`?

```js
{
	onlyAssignToVariable: true,
	noAssignment: false, // Default fasle, but can forbid
	noDefaultValueInAssign: true,
	noComputedProperty: true,
	maximumLevel: 2,
}
```

**Comment by brettz9** [comment]
Looks good to me, but what do you mean `noAssignment`--you mean no destructuring at all?

Also, maybe something like `noRename` for your other case?

**Comment by brettz9** [comment]
Btw, I already have a rule implemented for `onlyAssignToVariable` (if you mean no property references) which could serve as a base.

**Comment by fisker** [comment]
> what do you mean noAssignment--you mean no destructuring at all?

Yes, because assignment from object needs parens, I find many people don't like it.

**Comment by fisker** [comment]
If we add `noRename`, it should allow some excptions, eg: keywords, exisiting variables.

```js
const {
	a,
	b,
	c,
	default: d,
	v: e,
} = v;
```

**Comment by brettz9** [comment]
> > what do you mean noAssignment--you mean no destructuring at all?
> 
> Yes, because assignment from object needs parens, I find many people don't like it.

I see. I think [prefer-destructuring](https://eslint.org/docs/rules/prefer-destructuring) can effectively do this though with the option `AssignmentExpression` (as opposed to `VariableDeclarator`).
